designs/0034-jacobian.md

```diff
@@ -0,0 +1,320 @@
+- Feature Name: `jacobian` target
+- Start Date: 01-25-2024
+
+# Summary
+[summary]: #summary
+
+This design doc proposes adding a `jacobian` target and a new block for user defined constraints. The `jacobian` target will be accessible directly in `transformed parameters` and a new `constraints` block where users can define custom constraints similar to Stan's already existing constrained data types. The below examples shows an example use case.
+
+```stan
+constraints {
+ upper_bound {
+   real constrain(real x, real upper_bound) {
+     jacobian += x;
+     return upper_bound - exp(x);
+   }
+   real constrain(vector x, real upper_bound) {
+     jacobian += sum(x);
+     return upper_bound - exp(x);
+   }
+   real unconstrain(real x, real upper_bound) {
+     return log(upper_bound - x);
+   }
+   // Used for data
+   int validate(real x, real upper_bound) {
+     // For most this will be
+     return x < upper_bound;
+   }
+ }
+}
+data {
+ real lb;
+ real ub;
+ int N;
+}
+parameters {
+ //(2) User defined constraint
+ real<udc=upper_bound(ub)> b;
+ vector<udc=upper_bound(ub)>[N] b_vec;
+ real c_raw;
+}
+transformed parameters {
+ // (3) Transform and accumulates like stan math directly
+ real c = exp(c_raw) + lb;
+ // Underneath the hood, only actually calculated if
+ // Jacobian bool template parameter is set to true
+ jacobian += c_raw;
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+Given a function mapping $c$ that transforms a set of parameters from the unconstrained space $Y$ to the constrained space $X$, probability density function $\pi$, and a Jacobian determinant function over the constrained space $J(c(y))$, Stan calculates the log transformed density function [2]
```

Comment:
Yes I think this is clearer. I added all of this. I included the section on choosing the "Jacobian" function, though I think the part about nonbijective functions is better for the docs instead of here. The current API should be able to handle the one to many like functions since we have the unit vector transform